File: import/new/1504.java
// https://leetcode.com/problems/count-submatrices-with-all-ones/
class Solution {
    public int numSubmat(int[][] mat) {
        int n = mat.length, m = mat[0].length, res = 0;
        int[] height = new int[m], stack = new int[m], sum = new int[m];
        for (int i = 0; i != n; i++) {
            for (int j = 0; j != m; j++)
                height[j] = mat[i][j] == 0 ? 0 : height[j] + 1;
            int top = -1;
            for (int j = 0; j < m; j++) {
                int hj = height[j];
                while (top >= 0 && height[stack[top]] >= hj)
                    top--;
                if (top == -1)
                    sum[j] = hj * (j + 1);
                else
                    sum[j] = sum[stack[top]] + hj * (j - stack[top]);
                res += sum[j];
                stack[++top] = j;
            }
        }
        return res;
    }
}

File: code/python/1504.py
# https://leetcode.com/problems/count-submatrices-with-all-ones/
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        heights = [0] * len(mat[0])
        res = 0
        for row in mat:
            S, stack = 0, []
            for i, cell in enumerate(row):
                if cell == 0:
                    S = heights[i] = 0
                    stack.clear()
                    continue
                cur_height = heights[i] + 1
                width = 0
                while stack and stack[-1][0] >= cur_height:
                    h, w = stack.pop()
                    S -= h * w
                    width += w
                width += 1
                heights[i] = cur_height
                stack.append((cur_height, width))
                S += cur_height * width
                res += S
        return res
